### PR TITLE
Remove dependency to one specific port in a client test

### PR DESCRIPTION
**Pull request by sancar**
Related test
ClientDiscoverySpiTest#testClientCanConnect_afterDiscoveryStrategyThrowsException

When the configured port is occupied because of another test, this
test was not working correctly. Made the client config so that it can
adapt to a different port.

fixes https://github.com/hazelcast/hazelcast/issues/17457

### CODE REVIEW

**Review by mdumandag — APPROVED** [review]

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2195/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2195/Hazelcast-pr-builder-2195.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MigrationCorrectnessTest>AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withRestart:126->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:221->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:225->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->PartitionCorrectnessTestSupport.access$000:66->PartitionCorrectnessTestSupport.assertSizeAndData:271->PartitionCorrectnessTestSupport.assertReplicaVersionsAndBackupValues:329 Versions null on [127.0.0.1]:5704, partitionId: 9
[INFO] 
[ERROR] Tests run: 31485, Failures: 1, Errors: 0, Skipped: 977
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientDiscoverySpiTest.java [hunk 1]
```diff
@@ -490,7 +491,11 @@ public PartitionGroupStrategy getPartitionGroupStrategy() {
 
     public static class ExceptionThrowingDiscoveryStrategyFactory implements DiscoveryStrategyFactory {
 
-        public ExceptionThrowingDiscoveryStrategyFactory() {
+        private final Address address;
+
+        public ExceptionThrowingDiscoveryStrategyFactory(Address address) {
+
```
Comment (puzpuzpuz) [inline hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientDiscoverySpiTest.java:497]: nit: redundant new line here

**Review by puzpuzpuz — APPROVED** [review]

**Comment by sancar** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2231/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2231/Hazelcast-pr-builder-2231.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientReliableTopicOnClusterRestartTest.shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant:185->lambda$shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant$5:187
[INFO] 
[ERROR] Tests run: 31480, Failures: 1, Errors: 0, Skipped: 977
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by sancar** [comment]
run-lab-run

**Pull request merged**